service/inject/inject/src/main/java/io/helidon/service/inject/InjectionMain.java

```diff
@@ -0,0 +1,255 @@
+     *     <li>{@link #afterServiceDescriptors(io.helidon.service.inject.InjectConfig.Builder)} to update the builder</li>
+     *     <li>{@link #init(InjectConfig)} to initialize the service registry</li>
+     * </ol>
+     *
+     * @param arguments command line arguments
+     */
+    protected void start(String[] arguments) {
+        var config = configBuilder(arguments);
+        beforeServiceDescriptors(config);
+        serviceDescriptors(config);
+        afterServiceDescriptors(config);
+        var registry = init(config.build());
+        afterInit(registry);
+    }
+
+    /**
+     * Method that registers all service descriptors.
+     * This method is expected to be code generated.
+     *
+     * @param configBuilder in progress config builder
+     */
+    protected abstract void serviceDescriptors(InjectConfig.Builder configBuilder);
+
+    /**
+     * Create a config builder from command line arguments.
+     * This method will create default configuration (but not register it as a global config, so we still use the registry
+     * to create a config instance), configure the {@link io.helidon.service.inject.InjectConfig.Builder} from it,
+     * then disable discovery of services.
+     *
+     * @param arguments command line arguments
+     * @return a new config builder
+     */
+    protected InjectConfig.Builder configBuilder(String[] arguments) {
+        Config config;
+        if (GlobalConfig.configured()) {
+            config = GlobalConfig.config();
+        } else {
+            config = Config.create();
+        }
+
+        return InjectConfig.builder()
+                .config(config.get("registry"))
+                .discoverServices(discoverServices())
+                .discoverServicesFromServiceLoader(discoverServices());
+    }
+
+    /**
+     * Whether to discover services from classpath and from service loader.
+     * Defaults to {@code false}, should be overridden when
+     * {@link #serviceDescriptors(io.helidon.service.inject.InjectConfig.Builder)}
+     * does not configure all services.
+     *
+     * @return whether to discover services
+     */
+    protected boolean discoverServices() {
+        return false;
+    }
+
+    /**
+     * Called before service descriptors are configured in the generated main class.
+     * This method is invoked from generated {@link #start(String[])}.
+     *
+     * @param configBuilder in-progress config builder
+     */
+    protected void beforeServiceDescriptors(InjectConfig.Builder configBuilder) {
+    }
+
+    /**
+     * Called after service descriptors are configured in the generated main class, before the registry is initialized.
+     * This method is invoked from generated {@link #start(String[])}
+     *
+     * @param configBuilder in-progress config builder
+     */
+    protected void afterServiceDescriptors(InjectConfig.Builder configBuilder) {
+    }
```

Comment:
Because you are not required to implement them. The only one that is code generated is `serviceDescriptors`, all other methods are for extensibility by the user.